File: examples/eskf_demo.py
"""
End-to-end GNSS+IMU loose-coupling fusion demo using ESKF.

Pipeline:
  1. Generate a ground-truth trajectory (constant speed + turn).
  2. Synthesize noisy IMU (tactical grade) and GNSS (automotive grade).
  3. Run ESKF: IMU mechanization at 100 Hz, GNSS position+velocity updates at 5 Hz.
  4. Compare:
     - pure IMU dead-reckoning (drifts without updates)
     - ESKF fused solution vs ground truth
  5. Report position/velocity/attitude errors, bias estimation quality.

Usage:  python examples/eskf_demo.py [--seed N]
"""

import argparse
import time

import numpy as np

import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from sensor_sim.trajectory import Trajectory, Waypoint
from sensor_sim.imu import IMUSensor, SensorGrade
from sensor_sim.gnss import GNSSSensor, GNSSGrade
from sensor_sim.eskf import ESKF, ESKFConfig
from sensor_sim.utils import quat_to_euler


def build_trajectory(dt: float) -> Trajectory:
    """Constant-speed 8-shape-ish trajectory with a turn, 60 s."""
    wp = [
        Waypoint(t=0.0,  pos=np.array([0, 0, 0]),      vel=np.array([10, 0, 0]),
                 att=np.array([1, 0, 0, 0]),           omega=np.array([0, 0, 0])),
        Waypoint(t=15.0, pos=np.array([150, 0, 0]),    vel=np.array([10, 0, 0]),
                 att=np.array([1, 0, 0, 0]),           omega=np.array([0, 0, 0])),
        # Right turn: 90 deg over 5 s at 10 m/s (arc radius ~31.8 m)
        Waypoint(t=20.0, pos=np.array([150, 50, 0]),   vel=np.array([0, 10, 0]),
                 att=np.array([np.cos(np.pi/4), 0, 0, np.sin(np.pi/4)]),
                 omega=np.array([0, 0, np.pi/10])),
        Waypoint(t=40.0, pos=np.array([150, 250, 0]),  vel=np.array([0, 10, 0]),
                 att=np.array([np.cos(np.pi/4), 0, 0, np.sin(np.pi/4)]),
                 omega=np.array([0, 0, 0])),
        # Left turn back
        Waypoint(t=50.0, pos=np.array([50, 250, 0]),   vel=np.array([-10, 0, 0]),
                 att=np.array([np.cos(np.pi/4), 0, 0, -np.sin(np.pi/4)]),
                 omega=np.array([0, 0, -np.pi/10])),
        Waypoint(t=60.0, pos=np.array([-50, 250, 0]),  vel=np.array([-10, 0, 0]),
                 att=np.array([1, 0, 0, 0]),           omega=np.array([0, 0, 0])),
    ]
    return Trajectory(wp, dt=dt)


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--seed", type=int, default=42)
    args = ap.parse_args()

    dt_imu = 0.01        # 100 Hz
    dt_gnss = 0.2        # 5 Hz (GNSS sensor output rate)
    T = 60.0

    traj = build_trajectory(dt_imu)
    imu = IMUSensor(SensorGrade.TACTICAL, dt=dt_imu, seed=args.seed)
    gnss = GNSSSensor(GNSSGrade.AUTOMOTIVE, dt=dt_imu, seed=args.seed + 1)

    # --- ESKF setup ---------------------------------------------------
    # Tune measurement noise to match the automotive-grade sensor:
    # pos noise 1 m + multipath 2 m (Gauss-Markov, tau=20 s) -> use ~2.2 m
    # vel noise 0.05 m/s -> use 0.1 m/s to stay conservative.
    cfg = ESKFConfig(
        acc_noise_density=1e-2,      # m/s^2/sqrt(Hz)
        gyr_noise_density=1e-3,      # rad/s/sqrt(Hz)
        acc_bias_rw=1e-4,
        gyr_bias_rw=1e-5,
        init_pos_std=5.0,
        init_att_std_deg=5.0,
        gnss_pos_std=2.2,            # automotive-grade ~2.2 m incl. multipath
        gnss_vel_std=0.1,
    )
    eskf = ESKF(cfg)

    pts = traj.points
    p0 = pts[0]
    eskf.set_initial_state(p0.t, p0.pos + np.array([3, -2, 1.0]), p0.vel,
                           p0.att)   # 3 m initial position error

    # Pure dead-reckoning copy for comparison
    dr_p = p0.pos.copy(); dr_v = p0.vel.copy(); dr_q = p0.att.copy()

    errs_pos = []; errs_vel = []; errs_yaw = []
    dr_errs = []
    bias_est = []; bias_true = []

    gnss_next = dt_gnss
    t_start = time.time()

    for pt in pts:
        t = pt.t
        # IMU measurement
        a_true = traj.body_accel(pt)
        w_true = pt.omega
        a_m, w_m = imu.measure(a_true, w_true)

        # ESKF predict
        eskf.predict(a_m, w_m, dt_imu)

        # Dead reckoning (same mechanization, no updates)
        dr_q = _q_integrate(dr_q, w_m, dt_imu)
        R = _q_to_R(dr_q)
        dr_a_w = R @ (a_m - np.zeros(3)) + np.array([0, 0, -9.80665])
        dr_v += dr_a_w * dt_imu
        dr_p += dr_v * dt_imu

        # GNSS update (internal output-rate control)
        pos_m, vel_m, valid = gnss.measure(pt.pos, pt.vel)
        if valid:
            eskf.update_gnss(pos_m, vel_m)

        # Record errors at 2 Hz
        if abs(t - round(t / 0.5) * 0.5) < dt_imu * 0.5:
            e_pos = np.linalg.norm(eskf.position - pt.pos)
            e_vel = np.linalg.norm(eskf.velocity - pt.vel)
            e_yaw = abs(quat_to_euler(eskf.quaternion)[2] - quat_to_euler(pt.att)[2])
            errs_pos.append((t, e_pos)); errs_vel.append((t, e_vel)); errs_yaw.append((t, e_yaw))
            dr_errs.append((t, np.linalg.norm(dr_p - pt.pos)))
            bias_est.append((t, eskf.estimated_biases()[0].copy()))
            bias_true.append((t, imu.get_errors()["bias_acc"]))

    t_el = time.time() - t_start

    # --- Report --------------------------------------------------------
    print(f"=== ESKF GNSS+IMU Loose Coupling (seed={args.seed}, {t_el:.1f}s sim) ===")
    t = np.array([x[0] for x in errs_pos])
    ep = np.array([x[1] for x in errs_pos])
    ev = np.array([x[1] for x in errs_vel])
    ey = np.array([x[1] for x in errs_yaw])
    dr = np.array([x[1] for x in dr_errs])

    print(f"\nIMU-only dead reckoning position error:  mean {dr.mean():6.2f} m  max {dr.max():6.2f} m")
    print(f"ESKF position error:                    mean {ep.mean():6.3f} m  max {ep.max():6.3f} m   (RMS {np.sqrt((ep**2).mean()):.3f} m)")
    print(f"ESKF velocity error:                    mean {ev.mean():6.3f} m/s max {ev.max():6.3f} m/s")
    print(f"ESKF yaw error:                         mean {np.rad2deg(ey.mean()):6.3f} deg  max {np.rad2deg(ey.max()):6.3f} deg")

    # Bias estimation convergence (last 10 s)
    bt = np.array([x[0] for x in bias_est]); be = np.array([x[1] for x in bias_est]); btrue = np.array([x[1] for x in bias_true])
    mask = bt > T - 10
    if mask.any():
        print(f"\nAccel bias estimate (last 10 s):  est {be[mask][-1].round(4)}  true {btrue[mask][-1].round(4)}")
        print(f"Gyro bias estimate:               est {eskf.estimated_biases()[1].round(5)}  "
              f"true {imu.get_errors()['bias_gyr'].round(5)}")
    return ep.max()


def _q_integrate(q, w, dt):
    """Quaternion integration (wxyz), w = body angular rate."""
    from sensor_sim.eskf import _quat_exp
    from sensor_sim.utils import quat_multiply
    qn = quat_multiply(q, _quat_exp(w * dt))
    return qn / np.linalg.norm(qn)


def _q_to_R(q):
    from sensor_sim.utils import quat_to_rotmat
    return quat_to_rotmat(q)


if __name__ == "__main__":
    main()


def main_rtk():
    """RTK-grade variant: shows the fusion accuracy ceiling."""
    import argparse
    ap = argparse.ArgumentParser()
    ap.add_argument("--seed", type=int, default=42)
    args = ap.parse_args()

    dt_imu = 0.01
    T = 60.0
    traj = build_trajectory(dt_imu)
    imu = IMUSensor(SensorGrade.TACTICAL, dt=dt_imu, seed=args.seed)
    gnss = GNSSSensor(GNSSGrade.RTK, dt=dt_imu, seed=args.seed + 1)

    cfg = ESKFConfig(
        acc_noise_density=1e-2, gyr_noise_density=1e-3,
        acc_bias_rw=1e-4, gyr_bias_rw=1e-5,
        init_pos_std=5.0, init_att_std_deg=5.0,
        gnss_pos_std=0.3, gnss_vel_std=0.05,
    )
    eskf = ESKF(cfg)
    pts = traj.points
    p0 = pts[0]
    eskf.set_initial_state(p0.t, p0.pos, p0.vel, p0.att)

    errs = []
    for pt in pts:
        a_m, w_m = imu.measure(traj.body_accel(pt), pt.omega)
        eskf.predict(a_m, w_m, dt_imu)
        pos_m, vel_m, valid = gnss.measure(pt.pos, pt.vel)
        if valid:
            eskf.update_gnss(pos_m, vel_m)
        if pt.t > 10.0:      # settled region only
            errs.append(np.linalg.norm(eskf.position - pt.pos))
    errs = np.array(errs)
    print(f"\n=== RTK variant (seed={args.seed}) ===")
    print(f"Settled position error (t>10s): RMS {np.sqrt((errs**2).mean()):.3f} m  max {errs.max():.3f} m")
    print(f"Updates: {eskf.state.n_updates}")


if __name__ == "__main__":
    main()

File: cpp/examples/eskf_demo.cpp
/**
 * End-to-end GNSS+IMU loose-coupling fusion demo — C++ port of
 * examples/eskf_demo.py (same 60 s trajectory, tactical IMU @ 100 Hz,
 * automotive GNSS @ 5 Hz, 3 m initial position error).
 *
 * Build & run from cpp/:  ./build/eskf_demo --seed 42
 * Reference (Python, seed 42):  ESKF pos RMS ~2.3 m, DR mean ~12 m,
 * yaw err mean ~0.2 deg.  The C++ port should land in the same band
 * (streams differ; see cpp/README.md).
 */

#include <chrono>
#include <cmath>
#include <cstdio>

#include "nav/eskf.hpp"
#include "nav/gnss.hpp"
#include "nav/imu.hpp"
#include "nav/trajectory.hpp"

using namespace nav;

static Trajectory buildTrajectory(double dt) {
  using V4 = Eigen::Vector4d;
  auto yawQuat = [](double yaw) { return V4(std::cos(yaw / 2), 0.0, 0.0, std::sin(yaw / 2)); };

  std::vector<Waypoint> wp = {
      {0.0, {0, 0, 0}, {10, 0, 0}, yawQuat(0), {0, 0, 0}},
      {15.0, {150, 0, 0}, {10, 0, 0}, yawQuat(0), {0, 0, 0}},
      // Right turn: 90 deg over 5 s at 10 m/s
      {20.0, {150, 50, 0}, {0, 10, 0}, yawQuat(M_PI / 2), {0, 0, M_PI / 10}},
      {40.0, {150, 250, 0}, {0, 10, 0}, yawQuat(M_PI / 2), {0, 0, 0}},
      // Left turn back
      {50.0, {50, 250, 0}, {-10, 0, 0}, yawQuat(M_PI), {0, 0, -M_PI / 10}},
      {60.0, {-50, 250, 0}, {-10, 0, 0}, yawQuat(2 * M_PI), {0, 0, 0}},
  };
  return Trajectory(wp, dt);
}

int main(int argc, char** argv) {
  uint64_t seed = 42;
  for (int i = 1; i + 1 < argc; ++i) {
    if (std::string(argv[i]) == "--seed") seed = std::stoul(argv[i + 1]);
  }
  const double dt_imu = 0.01;  // 100 Hz
  const double T = 60.0;

  Trajectory traj = buildTrajectory(dt_imu);
  ImuSensor imu(SensorGrade::kTactical, dt_imu, seed);
  GnssSensor gnss(GnssGrade::kAutomotive, dt_imu, seed + 1);

  // Match the Python demo tuning: automotive-grade GNSS -> 2.2 m, 0.1 m/s.
  EskfConfig cfg;
  cfg.acc_noise_density = 1e-2;
  cfg.gyr_noise_density = 1e-3;
  cfg.acc_bias_rw = 1e-4;
  cfg.gyr_bias_rw = 1e-5;
  cfg.init_pos_std = 5.0;
  cfg.init_att_std_deg = 5.0;
  cfg.gnss_pos_std = 2.2;
  cfg.gnss_vel_std = 0.1;
  Eskf eskf(cfg);

  const auto& pts = traj.points();
  const TrajPoint& p0 = pts.front();
  eskf.setInitialState(p0.t, p0.pos + Vec3(3, -2, 1.0), p0.vel, p0.att);

  // Pure dead-reckoning copy (no updates).
  Vec3 dr_p = p0.pos, dr_v = p0.vel;
  Eigen::Vector4d dr_q = p0.att;

  std::vector<double> t_rec, ep_rec, ev_rec, ey_rec, dr_rec;
  Vec3 ba_est_last = Vec3::Zero(), ba_true_last = Vec3::Zero();
  Vec3 bg_est = Vec3::Zero(), bg_true = Vec3::Zero();

  auto t_start = std::chrono::steady_clock::now();
  for (const auto& pt : pts) {
    Vec3 a_true = traj.bodyAccel(pt);
    Vec3 a_m, w_m;
    imu.measure(a_true, pt.omega, a_m, w_m);

    eskf.predict(a_m, w_m, dt_imu);

    // Dead reckoning: same mechanization, no GNSS.
    dr_q = quatNormalize(quatMultiply(dr_q, quatExp(w_m * dt_imu)));
    Vec3 dr_a_w = quatToRotmat(dr_q) * a_m + Vec3(0, 0, -kGravity);
    dr_v += dr_a_w * dt_imu;
    dr_p += dr_v * dt_imu;

    auto m = gnss.measure(pt.pos, pt.vel);
    if (m.valid) {
      eskf.updateGnss({*m.pos, m.vel, std::nullopt, std::nullopt});
    }

    // Record errors at 2 Hz.
    if (std::abs(pt.t - std::round(pt.t / 0.5) * 0.5) < dt_imu * 0.5) {
      t_rec.push_back(pt.t);
      ep_rec.push_back((eskf.position() - pt.pos).norm());
      ev_rec.push_back((eskf.velocity() - pt.vel).norm());
      double dyaw = quatToEuler(eskf.quaternion()).z() -
                    quatToEuler(pt.att).z();
      // Wrap to [-pi, pi] (the Python demo metric lacks this and spikes
      // to ~360 deg when either yaw crosses the +/-pi boundary).
      dyaw = std::atan2(std::sin(dyaw), std::cos(dyaw));
      ey_rec.push_back(std::abs(dyaw));
      dr_rec.push_back((dr_p - pt.pos).norm());
      ba_est_last = eskf.estimatedBiases().first;
      bg_est = eskf.estimatedBiases().second;
      ba_true_last = imu.trueAccBias();
      bg_true = imu.trueGyrBias();
    }
  }
  double t_el = std::chrono::duration<double>(std::chrono::steady_clock::now() - t_start).count();

  auto mean = [](const std::vector<double>& v) {
    double s = 0; for (double x : v) s += x; return s / v.size(); };
  auto rms = [&mean](const std::vector<double>& v) {
    double s = 0; for (double x : v) s += x * x; return std::sqrt(s / v.size()); };
  auto mx = [](const std::vector<double>& v) {
    return *std::max_element(v.begin(), v.end()); };

  std::printf("=== ESKF GNSS+IMU Loose Coupling (C++ port, seed=%llu, %.1fs sim) ===\n",
              static_cast<unsigned long long>(seed), t_el);
  std::printf("\nIMU-only dead reckoning position error:  mean %6.2f m  max %6.2f m\n",
              mean(dr_rec), mx(dr_rec));
  std::printf("ESKF position error:                    mean %6.3f m  max %6.3f m   (RMS %.3f m)\n",
              mean(ep_rec), mx(ep_rec), rms(ep_rec));
  std::printf("ESKF velocity error:                    mean %6.3f m/s max %6.3f m/s\n",
              mean(ev_rec), mx(ev_rec));
  std::printf("ESKF yaw error:                         mean %6.3f deg  max %6.3f deg\n",
              mean(ey_rec) * 180.0 / M_PI, mx(ey_rec) * 180.0 / M_PI);
  std::printf("\nAccel bias estimate (last 10 s):  est [%.4f %.4f %.4f]  true [%.4f %.4f %.4f]\n",
              ba_est_last.x(), ba_est_last.y(), ba_est_last.z(),
              ba_true_last.x(), ba_true_last.y(), ba_true_last.z());
  std::printf("Gyro bias estimate:               est [%.5f %.5f %.5f]  true [%.5f %.5f %.5f]\n",
              bg_est.x(), bg_est.y(), bg_est.z(), bg_true.x(), bg_true.y(), bg_true.z());
  std::printf("GNSS dropouts: %d (%.1f s)\n", gnss.stats().num_dropouts,
              gnss.stats().total_dropout_time_s);
  return 0;
}
